tests/black_compatibility_test.rs

```diff
@@ -0,0 +1,186 @@
+use std::path::{Path, PathBuf};
+use std::process::Stdio;
+use std::thread::sleep;
+use std::time::Duration;
+use std::{fs, process, str};
+
+use anyhow::{anyhow, Context, Result};
+use assert_cmd::{crate_name, Command};
+use itertools::Itertools;
+use ruff::checks::CheckCategory;
+use strum::IntoEnumIterator;
+use walkdir::WalkDir;
+
+/// Handles blackd process and allows submitting code to it for formatting.
+struct Blackd {
+    address: SocketAddr,
+    server: process::Child,
+    client: ureq::Agent,
+}
+
+impl Blackd {
+    pub fn new() -> Result<Self> {
+        // Get free TCP port to run on
+        let address =
+            TcpListener::bind(SocketAddrV4::new(Ipv4Addr::UNSPECIFIED, 0))?.local_addr()?;
+
+        let server = process::Command::new("blackd")
+            .args([
+                "--bind-host",
+                &address.ip().to_string(),
+                "--bind-port",
+                &address.port().to_string(),
+            ])
+            .stdout(Stdio::null())
+            .stderr(Stdio::null())
+            .spawn()
+            .context("Starting blackd")?;
+
+        // Wait for blackd to be ready
+        for _ in 0..20 {
+            match TcpStream::connect(address) {
+                Err(e) if e.kind() == ErrorKind::ConnectionRefused => {
+                    sleep(Duration::from_millis(50));
+                }
+                Err(e) => return Err(e.into()),
+                Ok(_) => break,
+            }
+        }
+
+        Ok(Self {
+            address,
+            server,
+            client: ureq::agent(),
+        })
+    }
+
+    /// Format given code with blackd.
+    pub fn check(&self, code: &[u8]) -> Result<Vec<u8>> {
+        match self
+            .client
+            .post(&format!("http://{}/", self.address))
+            .set("X-Line-Length", "88")
+            .send_bytes(code)
+        {
+            // 204 indicates the input wasn't changed during formatting, so
+            // we return the original.
+            Ok(response) => {
+                if response.status() == 204 {
+                    Ok(code.to_vec())
+                } else {
+                    let mut buf = vec![];
+                    response
+                        .into_reader()
+                        .take((1024 * 1024) as u64)
+                        .read_to_end(&mut buf)?;
+                    Ok(buf)
+                }
+            }
+            Err(ureq::Error::Status(_, response)) => Err(anyhow::anyhow!(
+                "Formatting with black failed: {}",
+                response.into_string()?
+            )),
+            Err(e) => Err(e.into()),
+        }
+    }
+}
+
+impl Drop for Blackd {
+    fn drop(&mut self) {
+        self.server.kill().expect("Couldn't end blackd process");
+    }
+}
+
+fn run_test(path: &Path, blackd: &Blackd, ruff_args: &[&str]) -> Result<()> {
+    let input = fs::read(path)?;
+
+    let step_1 = &Command::cargo_bin(crate_name!())?
```

Comment:
Do we need to invoke Ruff as an external process here?

Reply:
The initial ticket was mentioning integration test so I put it next to the others. As far as I understand that means I do not have access to the crate itself.

Nothing is really stopping us from moving this into the ruff tests though.